sycl/unittests/program_manager/CompileTarget.cpp

```diff
@@ -0,0 +1,249 @@
+    }
+    return -1;
+  };
+}
+
+class CompileTargetTest : public testing::Test {
+protected:
+  sycl::unittest::UrMock<> Mock;
+  CompileTargetTest() {
+    mock::getCallbacks().set_before_callback("urProgramCreateWithBinary",
+                                             &redefinedProgramCreateWithBinary);
+    mock::getCallbacks().set_after_callback("urDeviceGetInfo",
+                                            &redefinedDeviceGetInfo);
+    mock::getCallbacks().set_after_callback("urDeviceGet", &redefinedDeviceGet);
+  }
+};
+
+template <typename F>
+void checkUsedImageWithCompileTarget(const char *compile_target, F &&f) {
+  createWithBinaryLog.clear();
+  ASSERT_EQ(createWithBinaryLog.size(), 0U);
+  f();
+  ASSERT_EQ(createWithBinaryLog.size(), 1U);
+  EXPECT_EQ(createWithBinaryLog.back(), compile_target);
+}
+
+void launchSingleTaskKernel(queue q) {
+  q.single_task<SingleTaskKernel>([]() {});
+}
+
+TEST_F(CompileTargetTest, SingleTask) {
+  checkUsedImageWithCompileTarget("intel_gpu_skl", [&]() {
+    launchSingleTaskKernel(
+        queue{archSelector(syclex::architecture::intel_gpu_skl)});
+  });
+
+  checkUsedImageWithCompileTarget("intel_gpu_pvc", [&]() {
+    launchSingleTaskKernel(
+        queue{archSelector(syclex::architecture::intel_gpu_pvc)});
+  });
+
+  checkUsedImageWithCompileTarget("spir64_x86_64", [&]() {
+    launchSingleTaskKernel(queue{archSelector(syclex::architecture::x86_64)});
+  });
+}
+
+void launchNDRangeKernel(queue q) {
+  q.submit([&](handler &cgh) {
+    cgh.parallel_for<NDRangeKernel>(nd_range<1>(1, 1), [](auto) {});
+  });
+}
+
+TEST_F(CompileTargetTest, NDRangeKernel) {
+  checkUsedImageWithCompileTarget("intel_gpu_skl", [&]() {
+    launchNDRangeKernel(
+        queue{archSelector(syclex::architecture::intel_gpu_skl)});
+  });
+
+  checkUsedImageWithCompileTarget("intel_gpu_pvc", [&]() {
+    launchNDRangeKernel(
+        queue{archSelector(syclex::architecture::intel_gpu_pvc)});
+  });
+
+  checkUsedImageWithCompileTarget("spir64_x86_64", [&]() {
+    launchNDRangeKernel(queue{archSelector(syclex::architecture::x86_64)});
+  });
+}
+
+void launchRangeKernel(queue q) {
+  q.submit([&](handler &cgh) {
+    cgh.parallel_for<NDRangeKernel>(range<1>(1), [](auto) {});
+  });
+}
+
+TEST_F(CompileTargetTest, RangeKernel) {
+  checkUsedImageWithCompileTarget("intel_gpu_skl", [&]() {
+    launchRangeKernel(queue{archSelector(syclex::architecture::intel_gpu_skl)});
+  });
+
+  checkUsedImageWithCompileTarget("intel_gpu_pvc", [&]() {
+    launchRangeKernel(queue{archSelector(syclex::architecture::intel_gpu_pvc)});
+  });
+
+  checkUsedImageWithCompileTarget("spir64_x86_64", [&]() {
+    launchRangeKernel(queue{archSelector(syclex::architecture::x86_64)});
+  });
+}
+
+TEST_F(CompileTargetTest, NoDeviceKernel) {
+  try {
+    queue{}.single_task<NoDeviceKernel>([]() {});
+  } catch (sycl::exception &e) {
+    ASSERT_EQ(e.what(),
+              std::string("No kernel named NoDeviceKernel was found"));
```

Comment:
I wonder if this exception message should be improved, because it can be quite confusing. A kernel may be right in front of a developer, but it is not that we couldn't find it, it is our program is not compatible with a device because of how it was compiler.

I suppose that this error message is what we will see even without this patch, so I'm fine with addressing that separately.